### PR TITLE
Implement Clone::clone_from for Option and Result

### PR DESCRIPTION
**Pull request by m-ou-se**
See https://github.com/rust-lang/rust/issues/28481

### CODE REVIEW

File: src/libcore/option.rs [hunk 1]
```diff
@@ -1040,6 +1040,25 @@ fn expect_failed(msg: &str) -> ! {
 // Trait implementations
 /////////////////////////////////////////////////////////////////////////////
 
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<T: Clone> Clone for Option<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        match self {
+            &Some(ref x) => Some(x.clone()),
```
Comment (Centril) [inline src/libcore/option.rs]: Prefer using default match bindings (and in the other functions as well).

**Review by Centril — DISMISSED** [review]

**Comment by scottmcm** [comment]
These are still fully-structural `impl`s, right?  Could the `#[derive]` just be made to emit these?

**Comment by m-ou-se** [comment]
> These are still fully-structural `impl`s, right? Could the `#[derive]` just be made to emit these?

There's been a few discussions on this. https://github.com/rust-lang/rust/pull/27939 was closed because it would affect the compile times too much.

Also, I think it might go wrong when there are multiple fields: If some fields are succesfully clone_from'd, but another panics, the object might be left in an invalid state. But it should be possible if there's only one field.

File: src/libcore/option.rs (hunk 1 above)
Comment (m-ou-se) [reply src/libcore/option.rs]: Done

**Comment by Centril** [comment]
@rust-timer build 1406f733ce69b082ee0028996545b3c270625e95

**Comment by Centril** [comment]
Looks like noise.

**Comment by KodrAus** [comment]
Based on the discussion in #28481 this looks good to me.

@bors r+

**Pull request merged**